### DISCUSSION TITLE
Unable to access library store inside another react application

### DISCUSSION THREAD

**Discussion by anj-kr** [reduxjs/redux#4229] [Help]
Hello,

We are using a library where it handles authentication and the value of JWT is available in react library store. Now when we use the react library in another react application, we are not able to see the library store values(Only default value) is available in the react application.

I tried exporting the store directly with respect to react library but nothing seems to work.

My question is - Is it possible to use another react library store with updated value in any other react application?
Tried this as well - https://daveceddia.com/access-redux-store-outside-react/ but only default values are popping up with respect to react custom library and store. Any ideas or help on this please ? Thank you!

### What is the expected behavior?

Able to use the same store from react library and use it with another react application for fetching the user data.

**Comment by markerikson** [comment]
I'm afraid the question here is very unclear. "Can't see store values" and "nothing seems to work" aren't things we can help answer. Can you show an example project demonstrating the issue, preferably as a CodeSandbox.

**Reply by anj-kr** [reply]
Sorry for the lack of clarity. My question was how to use the redux store from the library inside react application. I created a react library project where it handles the authentication and the JWT tokens exist inside the library when it runs in another react application. Now how to use that store values outside the react library. When I tried accessing the library store, what I'm seeing is only the default value of the state, not the updated values. Not sure I can create an example quickly in sandbox. I need to create the whole library again with AuthGrant flow(OIDC) and use that library in another react example to show in the sandbox. Any other suggestions or help on this?

**Reply by markerikson** [reply]
I'm still not sure what you mean by "seeing the default value of the state".  Can you at least show a pseudocode example of what the setup is, what you expect to see, and what you're actually seeing?

Is it possible that you are ending up with two entirely separate Redux stores in the app, one from the library and one from the app itself?  Remember that only one Redux store can be provided to the application component tree, because that's what `<Provider>` passes down.

Most of the time a library should not be creating a Redux store itself.  Only one Redux store should exist, and it should be created by the application code.  A library should usually provide reducers or middleware that the app should add to its store on setup.

**Reply by anj-kr** [reply]
Thanks for the response again.  Yes, Problem here is there are 100 different applications that uses the same set of code to connect and receive the token(OIDC -AuthCode Grant). So in order to avoid this, we created a library with a store and it accepts the run time config for OIDC auth code flow and the library does the authentication and it holds the token in its own store. Now we want to use that store token in a component of another react application.

We have two stores in the app. One from the library and the other from its own. As you mentioned If every team who uses the library wants to implement the same store then it is a duplicate implementation right?

I understand now, the library should not have its own store. Is there any other pattern where we can access some function in library holding the token and can be used in react components outside the library.

**Reply by markerikson** [reply]
Can you show what this "library store" implementation actually looks like? What kinds of code does it involve - reducers, middleware, actions, ... ?

Typically, a library like this might provide a pre-built reducer, and the application code would have to add that reducer to the store at setup time.

**Reply by anj-kr** [reply]
Yes. The library store has the reducer, middleware and actions. For example :below the reducer
``
import {
  USER_SIGNED_OUT,
  STORE_USER_ERROR,
  USER_EXPIRED,
  STORE_USER,
  LOADING_USER
} from '../actions'

const initialState = {
  user: null,
  isLoadingUser: false
};

export default function (state = initialState, action) {
  switch (action.type) {
    case STORE_USER:
      return {
        ...state,
        isLoadingUser: false,
        user: action.payload
      }
    case LOADING_USER:
      return {
        ...state,
        isLoadingUser: true
      }
    case USER_EXPIRED:
    case STORE_USER_ERROR:
    case USER_SIGNED_OUT:
      return {
        ...state,
        user: null,
        isLoadingUser: false
      }
    default:
      return state
  }
}
``

I tried adding the library reducer as well in the react application store with other reducers, but that reducer is having the default state whenever accessing from the application components but not the updated one with the authenticated token. Am i missing anything like exporting the reducer or anything i'm not aware of ? please let me know.

**Reply by markerikson** [reply]
Honestly, this is what the Redux DevTools are for. What actions are being dispatched? Is the state being updated? If the reducer was actually added to the store, _and_ the correct actions are being dispatched, it _should_ be getting updated.  Where and how are you seeing "the default state" still?

**Reply by anj-kr** [reply]
Actions here are dispatched based on the authentication redirect mechanism for OIDC which exists inside the library .  Just to test further, when I added the state console in the library component updated state is available. But when I added the library reducer to the react application, I'm seeing only the default state in react application. Not sure I'm missing something here. Do I need to dispatch action from react application , that action is only for the library if the authentication redirect(Authorization Code Grant) call is successful?

**Reply by markerikson** [reply]
@anj-kr I'm afraid we're going in circles here, and you're really not answering my questions or providing any info I could actually use to help debug what's going on :(

There's nothing "magic" about this. It's all [core Redux fundamentals](https://redux.js.org/tutorials/fundamentals/part-3-state-actions-reducers):

- _If_ the reducer is added to the store, it should be running in response to dispatched actions
- _If_ the reducer sees a specific action that it cares about, it should be calculating a new state and returning a value
- You _should_ be able to see all dispatched actions in the Redux DevTools, and see exactly what state was updated for each action dispatched

So far you haven't shown me how the store is being configured, you haven't shown me any of the actions, you haven't shown me where and how you're trying to read the state, and you haven't shown me any examples of what the action log in the Redux DevTools look like. I simply have no way of knowing what your app is actually doing without seeing more code and seeing more information.

So. _Please_ take the time to investigate this further as I've asked. Look at the Redux DevTools.  See what actions are being dispatched. Throw console logging in the reducer and confirm it's running. This is, tbh, [basic debugging principles](https://blog.isquaredsoftware.com/2021/06/presentations-debugging-software/) here.

**Reply by anj-kr** [reply]
Sure. Thank you! I will take a look and update here again.
